lib/WUI/nhttp/server.cpp

```diff
@@ -0,0 +1,555 @@
+    if (is_active_slot(slot)) {
+        static_cast<Slot *>(slot)->release();
+    }
+}
+
+err_t Server::idle_conn_wrap(void *slot, altcp_pcb *conn) {
+    /*
+     * LwIP seems to be invoking the poll timer from time to time, not when
+     * inactive as documented. So track if we are or are not active on our side.
+     */
+    BaseSlot *s = static_cast<BaseSlot *>(slot);
+    if (!s->timeout.past()) {
+        return ERR_OK;
+    }
+
+    bool send_goodbye = (is_active_slot(slot) && static_cast<Slot *>(slot)->want_read());
+    lost_conn_wrap(slot, ERR_OK);
+    if (conn != nullptr) {
+        if (send_goodbye) {
+            /*
+             * This is a bit best-effort. We don't stress over this failing or whatever.
+             */
+            static const char goodbye[] = "HTTP/1.1 408 Request Timeout\r\n\r\n";
+            altcp_write(conn, goodbye, sizeof(goodbye), 0);
+            altcp_output(conn);
+        }
+        /*
+         * We don't know if connection-in-close can still generate callbacks
+         * but we sure don't want them.
+         */
+        remove_callbacks(conn);
+        if (altcp_close(conn) != ERR_OK) {
+            altcp_abort(conn);
+            return ERR_ABRT;
+        }
+    }
+
+    return ERR_OK;
+}
+
+err_t Server::received_wrap(void *raw_slot, struct altcp_pcb *conn, pbuf *data, err_t err) {
+    assert(raw_slot != nullptr);
+    assert(conn != nullptr);
+
+    BaseSlot *base_slot = static_cast<BaseSlot *>(raw_slot);
+    if (!is_active_slot(base_slot)) {
+        if (data == nullptr) {
+            // Closing an idle connection.
+            remove_callbacks(conn);
+            if (altcp_close(conn) == ERR_OK) {
+                return ERR_OK;
+            } else {
+                altcp_abort(conn);
+                return ERR_ABRT;
+            }
+        }
+        /*
+         * The connection was not yet active. Find a slot for it and activate it.
+         */
+        if (Slot *active_slot = base_slot->server->find_empty_slot(); active_slot != nullptr) {
+            assert(!active_slot->partial);
+            assert(active_slot->partial_consumed == 0);
+            assert(!active_slot->buffer);
+            assert(holds_alternative<Idle>(active_slot->state));
+
+            active_slot->state.emplace<handler::RequestParser>(*active_slot->server);
+            active_slot->conn = conn;
+            altcp_arg(conn, active_slot);
+            base_slot = active_slot;
+            // Activity set below
+
+            altcp_setprio(conn, ACTIVE_PRIO);
+        } else {
+            /*
+             * No slot available. Refuse the pbuf now (it'll be given to us
+             * later on, at which point we might have some resources
+             * available).
+             *
+             * Yes, we do _not_ free the pbuf.
+             */
+            return ERR_MEM;
+        }
+    }
+
+    assert(is_active_slot(base_slot));
+    Slot *slot = static_cast<Slot *>(base_slot);
```

Comment:
this looks suspicious ... a `dynamic_cast` should be a more appropriate construct at this spot (downcasting `Server::BaseSlot*` to `Server::Slot*`)

Reply:
That won't bring any advantage here. Dynamic cast uses vtables to check which type it is and this one doesn't have a vtable ‒ no virtual tables. Or, at least that's the theory I know.

Reply:
ok